### PR TITLE
Introduce Coprocessor Storage trait

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR adds a Coprocessor `Storage` trait and implemented a `RangesScanner` based on this trait. `RangesScanner` is able to scan multiple ranges and providing a simple `next()` interface.

Also changed existing scan executors to utilize `RangesScanner`. And in order to test the `RangesScanner`, a `FixtureStorage` is also provided, which implements `Storage`.

Extracted from https://github.com/tikv/tikv/pull/5119 . The difference is in this PR current executors are still relying on `tikv::storage`. The next step is to make current executors totally rely on the `Storage` trait.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

New unit tests.

### CODE REVIEW

File: tests/integrations/coprocessor/test_select.rs [hunk 1]
```diff
@@ -107,7 +107,7 @@ fn test_stream_batch_row_limit() {
     let req = DAGSelect::from(&product).build();
     assert_eq!(req.get_ranges().len(), 1);
 
-    let mut expected_ranges_last_byte = vec![(0, 3), (3, 6), (6, 255)];
+    let mut expected_ranges_last_byte = vec![(0, 3), (3, 6), (6, 9)];
```
Comment (breeswish) [inline tests/integrations/coprocessor/test_select.rs]: This change may be not necessary any more. However my local environment is broken now.... will be updated later if necessary.

File: src/coprocessor/dag/executor/mod.rs [hunk 2]
```diff
@@ -498,6 +490,67 @@ pub mod tests {
 
         let (snapshot, start_ts) = test_store.get_snapshot();
         let store = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
-        Box::new(TableScanExecutor::table_scan(table_scan, key_ranges, store).unwrap())
+        Box::new(TableScanExecutor::table_scan(table_scan, key_ranges, store, false).unwrap())
+    }
+
+    pub struct TableData {
```
Comment (breeswish) [inline src/coprocessor/dag/executor/mod.rs]: Note: these are almost simply moved from the original `scanner.rs`

File: src/coprocessor/dag/storage/ranges_iter.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (breeswish) [inline src/coprocessor/dag/storage/ranges_iter.rs]: Note: this file is almost simply moved from the original `ranges_iter.rs`. However this new file no longer distinguish point range and interval range.

File: src/coprocessor/dag/storage/scanner.rs [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    current_range_upper_key: Vec<u8>,
+    working_range_begin_key: Vec<u8>,
+    working_range_end_key: Vec<u8>,
+}
+
+impl<T: Storage> RangesScanner<T> {
+    pub fn new(
+        storage: T,
+        ranges: Vec<Range>,
+        scan_backward_in_range: bool, // TODO: This can be const generics
```
Comment (lonng) [inline src/coprocessor/dag/storage/scanner.rs]: Can we define a struct and assign a name to each boolean `(false, false, false) / (true, true, false), ...blabla` is really hard to remember the meaning of each paramater)?

File: src/coprocessor/dag/storage/scanner.rs [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+                    }
+                    return Ok(None); // drained
+                }
+            };
+            if self.is_scanned_range_aware {
+                self.update_scanned_range_from_scanned_row(&some_row);
+            }
+            if some_row.is_some() {
+                // Retrieved one row from point range or interval range.
+                if let Some(r) = self.scanned_rows_per_range.last_mut() {
```
Comment (lonng) [inline src/coprocessor/dag/storage/scanner.rs]: It's better to move `scanned_rows_per_range` related logic to `RangesIterator`.

Comment (breeswish) [reply src/coprocessor/dag/storage/scanner.rs]: RangesIterator should only iterate ranges

File: src/coprocessor/dag/storage/scanner.rs (hunk 4 above)
Comment (breeswish) [reply src/coprocessor/dag/storage/scanner.rs]: Once we use const generics it should be a problem any more.

Comment (lonng) [reply src/coprocessor/dag/storage/scanner.rs]: There is a long time gap to wait for the const generics GA.

File: tests/integrations/coprocessor/test_select.rs [hunk 6]
```diff
@@ -107,13 +107,18 @@ fn test_stream_batch_row_limit() {
     let req = DAGSelect::from(&product).build();
     assert_eq!(req.get_ranges().len(), 1);
 
-    let mut expected_ranges_last_byte = vec![(0, 3), (3, 6), (6, 9)];
+    let ignored_suffix_len = 18; // tXXXXXXXX_rXXXXXXX (only ignore first 7 bytes of the row id)
```
Comment (lonng) [inline tests/integrations/coprocessor/test_select.rs]: It's better to use the constants in `coprocessor::codec::table` module.

File: src/coprocessor/dag/storage/fixture.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        Self {
+            data: Arc::new(data),
+            data_view_unsafe: None,
+            is_backward_scan: false,
+            is_key_only: false,
+        }
+    }
+}
+
+impl From<Vec<(Vec<u8>, Vec<u8>)>> for FixtureStorage {
```
Comment (lonng) [inline src/coprocessor/dag/storage/fixture.rs]: Do we need to implement the `From<<Vec<(&[u8], &[8])>>>` for `FixtureStorage`? (I think it's necessary if elsewhere will construct a `FixtureStorage` from `From<<Vec<(&[u8], &[8])>>>` like: https://github.com/tikv/tikv/pull/5129/files#diff-8b90a8d482d4d26d1843a4c39a31868bR228)

File: src/coprocessor/dag/storage/scanner.rs [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+    scan_backward_in_range: bool,
+    is_key_only: bool,
+
+    scanned_rows_per_range: Vec<usize>,
+
+    // The following fields are only used for calculating scanned range. Scanned range is only
+    // useful in streaming mode, where the client need to know the underlying physical data range
+    // of each response slice, so that partial retry can be non-overlapping.
+    is_scanned_range_aware: bool,
+    current_range_lower_key: Vec<u8>,
```
Comment (lonng) [inline src/coprocessor/dag/storage/scanner.rs]: I am curious why don't use `Range` instead of two individual fields to represent it (any consideration?).

Comment (breeswish) [reply src/coprocessor/dag/storage/scanner.rs]: Good catch

File: src/coprocessor/dag/storage/fixture.rs (hunk 7 above)
Comment (breeswish) [reply src/coprocessor/dag/storage/fixture.rs]: Good catch

**Pull request merged**